File: rust/src/s0053.rs
pub struct Solution {}

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut sum = nums[0];
        let mut max_val = nums[0];
        for i in 1..nums.len() {
            sum = nums[i].max(sum + nums[i]);
            max_val = max_val.max(sum);
        }
        max_val
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn max_sub_array_test() {
        assert_eq!(Solution::max_sub_array(vec![-1]), -1);
        assert_eq!(Solution::max_sub_array(vec![-1, 0]), 0);
        assert_eq!(Solution::max_sub_array(vec![-8, -2]), -2);
        assert_eq!(
            Solution::max_sub_array(vec![-2, 1, -3, 4, -1, 2, 1, -5, 4]),
            6
        );
        assert_eq!(Solution::max_sub_array(vec![1, 2, 3, 4]), 10);
    }
}

File: cpp/s0053.cc
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
  int maxSubArray(vector<int> &nums) {
    int max_value = nums[0];
    for (int i = 0; i < (int)nums.size(); ++i) {
      int win_sum = 0;
      for (int j = i; j < (int)nums.size(); ++j) {
        win_sum += nums[j];
        max_value = max(win_sum, max_value);
      }
    }
    return max_value;
  }
  int maxSubArray_on(vector<int> &nums) {
    int max_val = nums[0];
    int sum = 0;
    for (int i = 0; i < (int)nums.size(); i++) {
      sum += nums[i];
      max_val = max(sum, max_val);
      if (sum < 0)
        sum = 0;
    }
    return max_val;
  }
  int maxSubArray_on2(vector<int> &nums) {
    int sum = nums[0];
    int max_val = nums[0];
    for (int i = 1; i < (int)nums.size(); ++i) {
      sum = max(nums[i], sum + nums[i]);
      max_val = max(max_val, sum);
    }
    return max_val;
  }
};

int main() {
  // std::vector<int> v = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
  std::vector<int> v = {5, 4, -1, 7, 8};

  Solution s;
  cout << s.maxSubArray(v) << '\n';

  return 0;
}



















class Solution {
public:
  int maxSubArray(vector<int>& nums) {
    int max_sum = nums[0];
    int tmp_sum = 0;
    for (const int &num: nums) {
      if ((tmp_sum + num) > 0) {
        tmp_sum = tmp_sum + num;
        max_sum = max(max_sum, tmp_sum);
      } else {
        tmp_sum = 0;
        max_sum = max(max_sum, num);
      }
    }
    return max_sum;
  }
};
